### DISCUSSION TITLE
LangChain discussion

### DISCUSSION THREAD

**Discussion by shanecowherd** [intitni/CopilotForXcode#276] [General]
https://www.youtube.com/watch?v=9AXP7tCI9PI

Just watched this video from TechLead where he describes how he is using LangChain to inject code/samples from a local folder into GPT.

I'd love to see the option to create vectors using the current project source folder, as well as a bunch of open source MIT licensed Swift projects and if possible, public documentation from Apple and/or framework authors.

If licenses permit, for example, I'd love to be able to query, "how can I add alamofire" and it could give me instructions using post 2021 examples.

As far as licenses go, we could leave it up to the user to make sure they respect the licenses.

Maybe an interface to just post a bunch of GitHub repo links and local source folders?

**Comment by intitni** [comment]
I am currently working on something like that. The first version will allow users to import files and websites. Please stay tuned for the 1.0.0 release.

**Reply by shanecowherd** [reply]
Awesome!

**Comment by intitni** [comment]
It's actually way more complicated than what those videos showed, especially when you want to deal with queries across multiple large documents. I believe I will need a lot of trial-and-error attempts to make the queries accurate enough, so I will pause the development of this feature and work on something easier for now.

**Reply by shanecowherd** [reply]
Agreed, and it gets expensive if you are passing a bunch of data to GPT4

**Comment by KieranConlon** [comment]
I did a test by 'following along' to the YouTube video, writing a small program using Python.
I got it to work and could have a 'conversation' about information across multiple PDF documents I provided, it was OK but it was still pretty basic.  
I can appreciate that making a local LangChain-type implementation would be a lot of work.

Screen shot example attached.
<img width="1365" alt="image" src="https://github.com/intitni/CopilotForXcode/assets/33759139/8a6473f0-d497-4fcb-9a36-649ae109d0b9">

**Reply by intitni** [reply]
I actually stopped the feature 2 weeks ago, so I forgot about the reasons why I stopped. 

I do have a RetrievalQA and RefineDocumentsChain implementation used for the `@web` scope which can answer questions on a web page (also available on the RetrievalQAChain playground page). But I found that it was too slow and expensive, so I tried to let ChatGPT decide if it needs more information after building answers from each chunk, but it never stopped at the right place. 

I think the reason was not really about the accuracy. But looking at the result from the vector store, the order of similar thunks is not always optimal (I don't really remember), I believe there are a lot of things we can do in document splitting and embedding to make it work better.
